src/realm.js

```diff
@@ -1260,16 +1261,84 @@ export class Realm {
+      if (allPathsAreAbrupt(c.consequent)) {
         Path.pushInverseAndRefine(c.joinCondition);
-        if (c.alternate instanceof PossiblyNormalCompletion) pushPathConditionsLeadingToNormalCompletion(c.alternate);
-      } else if (c.alternate instanceof AbruptCompletion) {
+        if (c.alternate instanceof PossiblyNormalCompletion || c.alternate instanceof ForkedAbruptCompletion)
+          pushPathConditionsLeadingToNormalCompletion(c.alternate);
+      } else if (allPathsAreAbrupt(c.alternate)) {
         Path.pushAndRefine(c.joinCondition);
-        if (c.consequent instanceof PossiblyNormalCompletion) pushPathConditionsLeadingToNormalCompletion(c.consequent);
+        if (c.consequent instanceof PossiblyNormalCompletion || c.consequent instanceof ForkedAbruptCompletion)
+          pushPathConditionsLeadingToNormalCompletion(c.consequent);
+      } else if (allPathsAreNormal(c.consequent)) {
+        if (!allPathsAreNormal(c.alternate)) {
+          let alternatePC = getNormalPathConditionFor(c.alternate);
+          let disjunct = AbstractValue.createFromLogicalOp(realm, "||", c.joinCondition, alternatePC);
+          Path.pushAndRefine(disjunct);
+        }
+      } else if (allPathsAreNormal(c.alternate)) {
+        let consequentPC = getNormalPathConditionFor(c.consequent);
+        let inverse = AbstractValue.createFromUnaryOp(realm, "!", c.joinCondition);
+        let disjunt = AbstractValue.createFromLogicalOp(realm, "||", inverse, consequentPC);
+        Path.pushAndRefine(disjunt);
+      } else {
+        let jc = c.joinCondition;
+        let consequentPC = AbstractValue.createFromLogicalOp(realm, "&&", jc, getNormalPathConditionFor(c.consequent));
+        let ijc = AbstractValue.createFromUnaryOp(realm, "!", jc);
+        let alternatePC = AbstractValue.createFromLogicalOp(realm, "&&", ijc, getNormalPathConditionFor(c.alternate));
+        let disjunt = AbstractValue.createFromLogicalOp(realm, "||", consequentPC, alternatePC);
+        Path.pushAndRefine(disjunt);
+      }
+    }
+
+    function allPathsAreAbrupt(c: Completion): boolean {
+      if (c instanceof PossiblyNormalCompletion || c instanceof ForkedAbruptCompletion)
+        return allPathsAreAbrupt(c.consequent) && allPathsAreAbrupt(c.alternate);
+      if (c instanceof AbruptCompletion) return !(c instanceof ErasedAbruptCompletion);
+      return false;
+    }
+
+    function allPathsAreNormal(c: Completion): boolean {
+      if (c instanceof PossiblyNormalCompletion || c instanceof ForkedAbruptCompletion)
+        return allPathsAreNormal(c.consequent) && allPathsAreNormal(c.alternate);
+      if (c instanceof AbruptCompletion) return c instanceof ErasedAbruptCompletion;
+      return true;
+    }
+
+    function getNormalPathConditionFor(c: Completion): Value {
+      invariant(c instanceof PossiblyNormalCompletion || c instanceof ForkedAbruptCompletion);
+      if (allPathsAreAbrupt(c.consequent)) {
```

Comment:
Maybe leave behind a comment that this is basically manual simplification, and the code could be simplified if we would trust the simplifier more.